internal/orchestrion/_integration/internal/agent/agent.go

```diff
@@ -6,167 +6,148 @@
-		return nil, err
-	}
-	session = &Session{agent: a, token: token}
-	if !a.currentSession.CompareAndSwap(nil, session) {
-		return nil, errors.New("a test session is already in progress")
+		return err
 	}
-	defer func() {
-		if err != nil {
-			a.currentSession.Store(nil)
-			session = nil
-		}
-	}()
+	defer req.Body.Close()
 
-	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
-	defer cancel()
+	_, err = dest.UnmarshalMsg(b)
+	return err
+}
 
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("http://127.0.0.1:%d/test/session/start?test_session_token=%s", a.port, session.token.String()), nil)
-	if err != nil {
-		return nil, err
-	}
+func New(t *testing.T) *MockAgent {
+	return &MockAgent{T: t}
+}
 
-	for {
-		resp, err := internalClient.Do(req)
-		if err != nil {
-			select {
-			case <-ctx.Done():
-				return nil, err
-			default:
-				time.Sleep(100 * time.Millisecond)
-				continue
-			}
-		}
-		if resp.StatusCode != 200 {
-			return nil, fmt.Errorf("test agent returned non-200 status code: HTTP %s", resp.Status)
-		}
-		break
-	}
+func (m *MockAgent) Start(t *testing.T) {
+	m.T.Log("mockagent: starting")
+
+	srv := httptest.NewServer(m)
+	m.srv = srv
+	t.Cleanup(srv.Close)
+
+	srvURL, err := url.Parse(srv.URL)
+	require.NoError(t, err)
 
-	t.Logf("Started test session with ID %s\n", session.token.String())
 	tracer.Start(
-		tracer.WithAgentAddr(fmt.Sprintf("127.0.0.1:%d", a.port)),
-		tracer.WithHTTPClient(&http.Client{Transport: session}),
+		tracer.WithAgentAddr(srvURL.Host),
+		tracer.WithHTTPClient(&internalClient),
 		tracer.WithSampler(tracer.NewAllSampler()),
 		tracer.WithLogStartup(false),
 		tracer.WithLogger(testLogger{t}),
 	)
 	t.Cleanup(tracer.Stop)
-
-	return session, nil
 }
 
-func (a *MockAgent) Close() error {
-	if !a.currentSession.CompareAndSwap(nil, nil) {
-		return errors.New("cannot close agent while a test session is in progress")
-	}
+func (m *MockAgent) Traces(t *testing.T) trace.Traces {
+	m.T.Log("mockagent: fetching spans")
 
-	a.processCancel()
-	if err := a.process.Wait(); err != nil {
-		return err
+	tracer.Flush()
+	tracer.Stop()
+	m.srv.Close()
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	spansByID := make(map[trace.ID]*trace.Trace)
+	for _, payload := range m.payloads {
+		for _, spans := range payload {
+			for _, span := range spans {
+				b, err := json.Marshal(span)
+				require.NoError(t, err)
+
+				var tr trace.Trace
+				err = json.Unmarshal(b, &tr)
+				require.NoError(t, err)
+
+				spansByID[trace.ID(span.SpanID)] = &tr
+			}
+		}
+	}
+	var result trace.Traces
+	for _, span := range spansByID {
+		if span.ParentID == 0 {
+			result = append(result, span)
+			continue
+		}
+		parent, ok := spansByID[span.ParentID]
+		if ok {
+			parent.Children = append(parent.Children, span)
+		}
 	}
+	return result
+}
 
-	return os.RemoveAll(a.virtualEnv)
+type testLogger struct {
+	*testing.T
 }
+
+func (l testLogger) Log(msg string) {
+	l.T.Log(msg)
+}
+
+var (
```

Comment:
nit: It's weird for me to see the package level variable at the bottom of a file 😅